docs/design/gateway/lwM2M-gateway.md

```diff
@@ -0,0 +1,181 @@
+
+Then, the gateway will handle the request from the server by calling deviceShifu instruction with the LwM2M Object and Resource.
+
+Before the gateway stop, it will deregister from the server and stop the LwM2M Client.
+
+```mermaid
+flowchart BT
+
+ls[LwM2M-server]
+
+subgraph EdgeNode
+    subgraph Shifu
+      subgraph ds1[deviceshifu-LwM2M]
+          dsh[deviceshifu-http]
+          gl1[LwM2M-gateway]
+          dsh <-->|HTTP| gl1
+      end
+      subgraph ds2[deviceshifu-MQTT]
+          dsm[deviceshifu-MQTT]
+          gl2[LwM2M-gateway]
+          dsm <-->|HTTP| gl2
+      end
+    end
+end
+
+
+dh[device-http]
+dm[device-mqtt]
+
+
+dh -->|HTTP| dsh
+gl1 -->|LwM2M| ls
+
+dm -->|MQTT| dsm
+gl2 -->|LwM2M| ls 
+```
+
+### What will the gateway do?
+
+1. Start the LwM2M Client and get all the device info from the deviceShifu.
+2. Register to the server and update the device info.
+3. handle the request from the server.
+4. When server enable Observe feature, the gateway will notify the server when the data changed or timeout.
+5. When server send the read or write request, the gateway will call the deviceShifu to get the data or set the data.
+6. Before gateway shutdown, deregister from the server and stop the LwM2M Client.
+7. When server disconnect, the gateway will try to reconnect to the server and register again.
+8. When call deviceShifu instruction timeout, the gateway will return the error message to the server.
+
+```mermaid
+sequenceDiagram
+   participant ds as DeviceShifu
+   participant gw as Gateway
+   participant s as Server
+
+   gw ->> s: Register
+   s ->> gw: Created
+
+   note over ds: read or write data
+   s ->> gw: Get Device Info
+   gw ->> ds: Get Device Info
+   ds ->> gw: Device Info
+   gw ->> s: return Device Info
+
+   note over ds: observe data
+   s ->> gw: observe object
+   gw ->> ds: Created
+   loop Get Device Data in a interval
+      gw ->> ds: Get Device Data
+      ds ->> gw: Get Device Data
+      alt data changed or timeout
+         gw ->> s: Notify Data
+      end
+   end
+```
+
+### Detail Design
+
+#### Read Request
+
+When the server send the read request to the gateway, the gateway will call the deviceShifu instruction with `GET` method. 
+The gateway will get the data from the deviceShifu and return the data to the server.
+
+#### Write Request
+
+When the server send the write request to the gateway, the gateway will call the deviceShifu instruction with `PUT` method with the data in the request body. and return with changed status code to server
+
+#### Execute Request
+
+When the server send the execute request to the gateway, the gateway will call the deviceShifu instruction with `POST` method without request body.
+
+#### Observe and Notify
+
+When the server enable the observe feature, the gateway will get the data from the deviceShifu in a interval and notify the server when the data changed or timeout.
```

Comment:
what happens if the deviceShifu is LwM2M and it enables Observe? does gateway get the data in an interval or on data change event?